### PR TITLE
API: Remove The MapIter API from public 

### PR DESCRIPTION
**Pull request by seberg**
This API is very slow to use compared to normal advanced indexing
(which reaches deep into the much richer internals).
Making this fast is difficult and the only known users is Theano:
* While probably still in use Theano is end-of-life for a long time.
* The use-case in Theano would be better served with `ufunc.at` or
  direct advanced indexing use probably (even if that requires
  wrapping into NumPy arrays).  Because it will be vastly faster
  normally.

So remove it (make it private) to allow making the ABI of it not be
a strange mix for backwards compatibility.

---

Revival of gh-24274, because I now have another reason.  Bumping maxdims to 64 cannot be done for the mapiter if it is public.  So just remove it, because nobody uses it (as in the other PR, the only public user and likely the reason it was made public initially gave a thumbs-up).

Otherwise, we would either not support 64dims in all advanced indexing paths or require odd hacks to keep practically unused API from working.

### CODE REVIEW

File: numpy/_core/src/multiarray/mapping.h [hunk 1]
```diff
@@ -4,6 +4,80 @@
 extern NPY_NO_EXPORT PyMappingMethods array_as_mapping;
 
 
+/*
+ * Store the information needed for fancy-indexing over an array. The
+ * fields are slightly unordered to keep consec, dataptr and subspace
+ * where they were originally.
```
Comment (mattip) [inline numpy/_core/src/multiarray/mapping.h:10]: Is there any need to reorder the fields?

File: doc/release/upcoming_changes/25138.c_api_removal.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+* The advanced indexing ``MapIter`` and related API has been removed.
+  The (truly) public part of it was not well tested and had only one
+  known user (Theano).  Making it private will simplify improvements
+  to speed up ``ufunc.at`` and make advanced indexing more maintainable
+  and was important for allowing 64 dimensional arrays.
+  Please let us know if this API is important to you so we can find a
+  solution together.
```
Comment (mattip) [inline doc/release/upcoming_changes/25138.c_api_removal.rst:7]: Does this blurb need a title line?

**Review by mattip — APPROVED** [review]
LGTM, pretty straightforward.

File: doc/release/upcoming_changes/25138.c_api_removal.rst (hunk 2 above)
Comment (mattip) [reply doc/release/upcoming_changes/25138.c_api_removal.rst:7]: No it [does not](https://output.circle-artifacts.com/output/job/c3dde236-55f3-4d85-a0e2-a18067d294ad/artifacts/0/doc/build/html/release/2.0.0-notes.html#numpy-2-0-c-api-removals)

**Comment by mattip** [comment]
Thanks @seberg

File: numpy/_core/src/multiarray/mapping.h (hunk 1 above)
Comment (seberg) [reply numpy/_core/src/multiarray/mapping.h:10]: Hmmm, will have a look.  I am not quite sure I found everything that can be cleaned out also.

**Pull request merged**